### ISSUE TITLE
Segfault when check-bounds=yes

### ISSUE THREAD

**Issue by m-j-w** [JuliaLang/julia#30996]
I believe to have run into an issue which is not related to my tinkering with llvmcall. When I run a minimum working example from the command line with `--check-bounds=yes`, and as `Pkg.test` would do, Julia segfaults. In fact, package tests failing and regular operations working fine were the reason I discovered this. The concerned package is [CpuId](https://github.com/m-j-w/CpuId.jl), of which I'm the author.

A minimum failing example:
```
function cpuinfo_mwe()
    CpuInstructions.cpuid_llvm(0x8000_0000,0x0000_0000)
    ([ [0] ], [ ])
end
```
When commenting out either the first or the second line, everything runs as expected. The segfault also disappears when reducing the result to another type rather than a `Tuple{Array{Array{Int64,1},1},Array{Any,1}}`. (although the Int64 is not relevant)

The cpuid_llvm function can be found here: (https://github.com/m-j-w/CpuId.jl/blob/c86b1d63e1e1f3d78e83975fdb38d4ce1b9f9af8/src/CpuInstructions.jl#L38)

**Edit:** When changing the `@inline cpuid_llvm(...)` to a `@noinline cpuid_llvm(...)` the segfault disappears, which is also my current workaround.

The above is packaged into a CpuId branch: `]add CpuId#CpuIdSegFault`.

It should then be reproducible with the command line
```
0 ❱❱ julia --check-bounds=yes -e "using CpuId ; CpuId.cpuinfo_mwe()"

signal (11): Segmentation fault
in expression starting at no file:0
setindex! at ./array.jl:767 [inlined]
vect at ./array.jl:130 [inlined]
cpuinfo_mwe at /home/mjw/.julia/dev/CpuId/src/CpuId.jl:1000
jl_fptr_trampoline at /buildworker/worker/package_linux64/build/src/gf.c:1864
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2219
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:323
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:411
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:362 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:773
jl_interpret_toplevel_thunk_callback at /buildworker/worker/package_linux64/build/src/interpreter.c:885
unknown function (ip: 0xfffffffffffffffe)
unknown function (ip: 0x7f6442872a8f)
unknown function (ip: (nil))
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:894
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:764
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:713
jl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:793
eval at ./boot.jl:328
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2219
exec_options at ./client.jl:243
_start at ./client.jl:436
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2219
unknown function (ip: 0x40191d)
unknown function (ip: 0x401523)
__libc_start_main at /build/glibc-Cl5G7W/glibc-2.23/csu/../csu/libc-start.c:291
unknown function (ip: 0x4015c4)
Allocations: 146706 (Pool: 146601; Big: 105); GC: 0
[1]    5309 segmentation fault (core dumped)  julia --check-bounds=yes -e "using CpuId ; CpuId.cpuinfo_mwe()"

0 ❱❱ julia --check-bounds=no -e "using CpuId ; println(CpuId.cpuinfo_mwe())"
(Array{Int64,1}[[0]], Any[])
```

Using the original function "cpuinfo()" shows the segfault to occur in the _inlined_ function "cpuid_llvm", which I believe to be not correct. I'm somewhat confident in that since I carefully compared the output of `@code_llvm` and `@code_native` for with and without bounds checks, with and without optimisation, and found no relevant difference in the code that relates to the `cpuid_llvm` call.

This issue appears in all versions since Julia 0.7 on Windows and Linux, afaik, also unfortunately demonstrated many times by AppVeyor and Travis.

```
Julia Version 1.1.0
Commit 80516ca202 (2019-01-21 21:24 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E3-1225 v5 @ 3.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, skylake)
```

**Comment by m-j-w** [comment]
Apparently, I could directly have given a more convenient piece of code to demonstrate right away...
So here it is in condensed form with output given further down:

```julia
using InteractiveUtils
using Base: llvmcall

@inline cpuid_llvm(leaf::UInt32, subleaf::UInt32) =
           llvmcall("""
               ; leaf = %0, subleaf = %1, %2 is some label
               ; call 'cpuid' with arguments loaded into registers EAX = leaf, ECX = subleaf
               %3 = tail call { i32, i32, i32, i32 } asm sideeffect "cpuid",
                   "={ax},={bx},={cx},={dx},{ax},{cx},~{dirflag},~{fpsr},~{flags}"
                   (i32 %0, i32 %1)
               ; retrieve the result values and convert to vector [4 x i32]
               %4 = extractvalue { i32, i32, i32, i32 } %3, 0
               %5 = extractvalue { i32, i32, i32, i32 } %3, 1
               %6 = extractvalue { i32, i32, i32, i32 } %3, 2
               %7 = extractvalue { i32, i32, i32, i32 } %3, 3
               ; return the values as a new tuple
               %8  = insertvalue [4 x i32] undef, i32 %4, 0
               %9  = insertvalue [4 x i32]   %8 , i32 %5, 1
               %10 = insertvalue [4 x i32]   %9 , i32 %6, 2
               %11 = insertvalue [4 x i32]  %10 , i32 %7, 3
               ret [4 x i32] %11
           """
           # llvmcall requires actual types, rather than the usual (...) tuple
           , NTuple{4,UInt32}, Tuple{UInt32,UInt32}
           , leaf, subleaf)

function cpuinfo_mwe()
           a = cpuid_llvm(0x8000_0000,0x0000_0000)
           ([ [0] ], [ ])
end
```

```julia
0 ❱❱ julia --check-bounds=yes
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.1.0 (2019-01-21)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

[...copy&paste...]

julia> @code_native cpuinfo_mwe()
        .text
; ┌ @ REPL[4]:2 within `cpuinfo_mwe'
        pushq   %rbp
        movq    %rsp, %rbp
        pushq   %r15
        pushq   %r14
        pushq   %r13
        pushq   %r12
        pushq   %rbx
        andq    $-32, %rsp
        subq    $64, %rsp
        movq    %rsp, %rbx
        vxorps  %xmm0, %xmm0, %xmm0
        vmovaps %ymm0, (%rbx)
        movq    %rsi, 32(%rbx)
        movabsq $jl_alloc_array_1d, %r15
        movq    %fs:0, %r14
; │┌ @ REPL[3]:1 within `cpuid_llvm'
        movq    $4, (%rbx)
        movq    -15552(%r14), %rax
        movq    %rax, 8(%rbx)
        movq    %rbx, %rax
        movq    %rax, -15552(%r14)
        movl    $2147483648, %eax       # imm = 0x80000000
        xorl    %ecx, %ecx
        cpuid
; │└
; │ @ REPL[4]:3 within `cpuinfo_mwe'
; │┌ @ array.jl:130 within `vect'
; ││┌ @ abstractarray.jl:617 within `similar' @ abstractarray.jl:618
; │││┌ @ boot.jl:419 within `Type' @ boot.jl:411 @ boot.jl:402
        movabsq $jl_system_image_data, %rdi
        movl    $1, %esi
        vzeroupper
        callq   *%r15
        movq    %rax, %r12
; ││└└
; ││┌ @ array.jl:767 within `setindex!'
        cmpq    $0, 8(%r12)
        je      L343
        movq    (%r12), %rax
        movq    $0, (%rax)
        movq    %r12, 16(%rbx)
; │└└
; │┌ @ boot.jl:402 within `vect'
        movabsq $jl_system_image_data, %rdi
        movl    $1, %esi
        callq   *%r15
        movq    %rax, %r15
; │└
; │┌ @ array.jl:130 within `vect'
; ││┌ @ array.jl:767 within `setindex!'
        cmpq    $0, 8(%r15)
        je      L385
        movzwl  16(%r15), %eax
        andl    $3, %eax
        movq    %r15, %rdi
        cmpl    $3, %eax
        jne     L195
        movq    40(%r15), %rdi
L195:
        addq    $-15552, %r14           # imm = 0xC340
; │││ @ array.jl:767 within `setindex!'
        movq    (%r15), %r13
        movl    -8(%rdi), %eax
        notl    %eax
        testb   $3, %al
        jne     L222
        testb   $1, -8(%r12)
        je      L329
L222:
        movq    %r12, (%r13)
        movq    %r15, 24(%rbx)
; ││└
; ││ @ array.jl:129 within `vect'
; ││┌ @ boot.jl:421 within `Type' @ boot.jl:402
        movabsq $140267440918592, %rdi  # imm = 0x7F928EFD5040
        xorl    %esi, %esi
        movabsq $jl_alloc_array_1d, %rax
        callq   *%rax
        movq    %rax, %r12
        movq    %r12, 16(%rbx)
; │└└
        movabsq $jl_gc_pool_alloc, %rax
        movl    $1688, %esi             # imm = 0x698
        movl    $32, %edx
        movq    %r14, %rdi
        callq   *%rax
        movabsq $140267478278928, %rcx  # imm = 0x7F9291376310
        movq    %rcx, -8(%rax)
        movq    %r15, (%rax)
        movq    %r12, 8(%rax)
        movq    8(%rbx), %rcx
        movq    %rcx, (%r14)
        leaq    -40(%rbp), %rsp
        popq    %rbx
        popq    %r12
        popq    %r13
        popq    %r14
        popq    %r15
        popq    %rbp
        retq
; │┌ @ array.jl:130 within `vect'
; ││┌ @ array.jl:767 within `setindex!'
L329:
        movabsq $jl_gc_queue_root, %rax
        callq   *%rax
        jmp     L222
L343:
        movq    %rsp, %rax
        leaq    -16(%rax), %rsi
        movq    %rsi, %rsp
        movq    $1, -16(%rax)
        movq    %r12, 16(%rbx)
        movabsq $jl_bounds_error_ints, %rax
        movl    $1, %edx
        movq    %r12, %rdi
        callq   *%rax
L385:
        movq    %rsp, %rax
        leaq    -16(%rax), %rsi
        movq    %rsi, %rsp
        movq    $1, -16(%rax)
        movq    %r15, 24(%rbx)
        movabsq $jl_bounds_error_ints, %rax
        movl    $1, %edx
        movq    %r15, %rdi
        callq   *%rax
; └└└

julia> cpuinfo_mwe()

signal (11): Segmentation fault
in expression starting at no file:0
setindex! at ./array.jl:767 [inlined]
vect at ./array.jl:130 [inlined]
cpuinfo_mwe at ./REPL[4]:3
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2219
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:323
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:411
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:362 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:773
jl_interpret_toplevel_thunk_callback at /buildworker/worker/package_linux64/build/src/interpreter.c:885
unknown function (ip: 0xfffffffffffffffe)
unknown function (ip: 0x7f9291895eff)
unknown function (ip: 0xffffffffffffffff)
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:894
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:764
jl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:793
eval at ./boot.jl:328
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2219
eval_user_input at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:85
macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:117 [inlined]
#26 at ./task.jl:259
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2219
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1571 [inlined]
start_task at /buildworker/worker/package_linux64/build/src/task.c:572
unknown function (ip: 0xffffffffffffffff)
Allocations: 1223431 (Pool: 1223126; Big: 305); GC: 1
[1]    16475 segmentation fault (core dumped)  julia --check-bounds=yes

```

**Comment by ViralBShah** [comment]
Is this still an issue?

**Comment by m-j-w** [comment]
Seems to be resolved, tested on Version 1.7.2.
